src/test/java/jssc/junit/rules/VirtualPortRule.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * GNU Lesser General Public License v3.0
+ * Copyright (C) 2019
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package jssc.junit.rules;
+
+import static java.util.Arrays.asList;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Random;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class VirtualPortRule implements TestRule {
+
+  private static final Logger LOG = LoggerFactory.getLogger(VirtualPortRule.class);
+
+  private static final ExecutorService executor = Executors.newCachedThreadPool();
+
+  private final List<Future<?>> processes = new ArrayList<Future<?>>();
+
+  private OutputStream outputStream;
+
+  private boolean isAvailable;
+
+  private final Random random = new SecureRandom();
+
+  private File virtualCom1;
+  private File virtualCom2;
+
+  public VirtualPortRule() {
+  }
+
+  @Override
+  public Statement apply(final Statement base, final Description description) {
+    final String osName = System.getProperty("os.name", "").toLowerCase(Locale.ENGLISH);
+    if (!osName.startsWith("linux")) {
```

Comment:
Is this just `LC_ALL=C` equivalent?  What part of this code is incompatible with other unixes?

Reply:
This rule will use different methods to prepare the unit test depending on the OS.
On Linux-Systems it will start `socat`, and on windows it might start something else someday. The idea of this "if" is that the method should be extensible for other operating systems.

Reply:
But it's not used, it should be removed.  Socat is available for MacOS.  We shouldn't be hard-coding Linux unless necessary.

Reply:
What about windows, *BSD, etc.? I'd be willing to change it to a `switch-case-block` and create a `os family` variable instead.

Reply:
All unixes potentially have `socat` available.  Windows doesn't have a ready free replacement that I'm aware of because serial ports actually are hardware devices (physical or virtual) requiring device driver signing, etc.

Reply:
>  change it to a switch-case-block and create a os family variable instead.

Agreed.  We can tackle this at a future time if deemed necessary.  For now, I'm assuming that `socat` is available everywhere except Windows.  

>  Socat is available for MacOS

Clarification, it's only available through 3rd party repositories such as macports or homebrew.  When testing without `socat` installed, deadlocks would occur, so I've improved the conditional to additionally execute `socat -V` prior to running the tests.